platforms/core-configuration/declarative-dsl-core/src/main/kotlin/org/gradle/internal/declarativedsl/serialization/SchemaSerialization.kt

```diff
@@ -66,7 +68,13 @@ object SchemaSerialization {
                 subclass(DataTypeInternal.DefaultBooleanDataType::class)
                 subclass(DataTypeInternal.DefaultNullType::class)
                 subclass(DataTypeInternal.DefaultUnitType::class)
+            }
```

Comment:
As far as I remember, kotlinx.serialization wants `DefaultDataClass` and `DefaultEnumClass` listed in the subtypes of `DataType` as well.